Can't build the qoi package with Python 3.12

Note numpy 1.26 is needed with Python 3.12 because distutils is removed.
```
Building wheels for collected packages: qoi
  Building wheel for qoi (pyproject.toml): started
  Building wheel for qoi (pyproject.toml): finished with status 'error'
  error: subprocess-exited-with-error

  Building wheel for qoi (pyproject.toml) did not run successfully.
  exit code: 1

  [50 lines of output]
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build\lib.win-amd64-cpython-312
  creating build\lib.win-amd64-cpython-312\qoi
  copying src\qoi\benchmark.py -> build\lib.win-amd64-cpython-312\qoi
  copying src\qoi\_version.py -> build\lib.win-amd64-cpython-312\qoi
  copying src\qoi\__init__.py -> build\lib.win-amd64-cpython-312\qoi
  copying src\qoi\qoi.pxd -> build\lib.win-amd64-cpython-312\qoi
  running build_ext
  building 'qoi.qoi' extension
  creating build\temp.win-amd64-cpython-312
  creating build\temp.win-amd64-cpython-312\Release
  creating build\temp.win-amd64-cpython-312\Release\src
  creating build\temp.win-amd64-cpython-312\Release\src\qoi
  "C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.37.32822\bin\HostX86\x64\cl.exe" /c /nologo /O2 /W3 /GL /DNDEBUG /MD -DQOI_MALLOC=PyMem_Malloc -DQOI_FREE=PyMem_Free -IC:\Users\Gudni\AppData\Local\Temp\pip-build-env-c92eykgi\overlay\Lib\site-packages\numpy\core\include -ID:\Prusa-Firmware-Buddy-Private\.venv\include -IC:\Users\Gudni\AppData\Local\Programs\Python\Python312\include -IC:\Users\Gudni\AppData\Local\Programs\Python\Python312\Include "-IC:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.37.32822\include" "-IC:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Auxiliary\VS\include" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.22621.0\\um" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.22621.0\\shared" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.22621.0\\winrt" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.22621.0\\cppwinrt" /Tcsrc/qoi/implementation.c /Fobuild\temp.win-amd64-cpython-312\Release\src/qoi/implementation.obj
  implementation.c
  C:\Users\Gudni\AppData\Local\Temp\pip-install-y6f4ym_p\qoi_7194dc5fc5904c6d98e093ad731ed32d\src\qoi\phoboslab_qoi/qoi.h(597): warning C4996: 'fopen': This function or variable may be unsafe. Consider using fopen_s instead. To disable deprecation, use _CRT_SECURE_NO_WARNINGS. See online help for details.
  C:\Users\Gudni\AppData\Local\Temp\pip-install-y6f4ym_p\qoi_7194dc5fc5904c6d98e093ad731ed32d\src\qoi\phoboslab_qoi/qoi.h(641): warning C4267: '=': conversion from 'size_t' to 'int', possible loss of data
  C:\Users\Gudni\AppData\Local\Temp\pip-install-y6f4ym_p\qoi_7194dc5fc5904c6d98e093ad731ed32d\src\qoi\phoboslab_qoi/qoi.h(619): warning C4996: 'fopen': This function or variable may be unsafe. Consider using fopen_s instead. To disable deprecation, use _CRT_SECURE_NO_WARNINGS. See online help for details.
  "C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.37.32822\bin\HostX86\x64\cl.exe" /c /nologo /O2 /W3 /GL /DNDEBUG /MD -DQOI_MALLOC=PyMem_Malloc -DQOI_FREE=PyMem_Free -IC:\Users\Gudni\AppData\Local\Temp\pip-build-env-c92eykgi\overlay\Lib\site-packages\numpy\core\include -ID:\Prusa-Firmware-Buddy-Private\.venv\include -IC:\Users\Gudni\AppData\Local\Programs\Python\Python312\include -IC:\Users\Gudni\AppData\Local\Programs\Python\Python312\Include "-IC:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Tools\MSVC\14.37.32822\include" "-IC:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC\Auxiliary\VS\include" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.22621.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.22621.0\\um" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.22621.0\\shared" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.22621.0\\winrt" "-IC:\Program Files (x86)\Windows Kits\10\\include\10.0.22621.0\\cppwinrt" /Tcsrc/qoi/qoi.c /Fobuild\temp.win-amd64-cpython-312\Release\src/qoi/qoi.obj
  qoi.c
  C:\Users\Gudni\AppData\Local\Temp\pip-build-env-c92eykgi\overlay\Lib\site-packages\numpy\core\include\numpy\npy_1_7_deprecated_api.h(14) : Warning Msg: Using deprecated NumPy API, disable it with #define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION
  src/qoi/qoi.c(13239): warning C4996: 'Py_OptimizeFlag': deprecated in 3.12
  src/qoi/qoi.c(21985): error C2039: 'curexc_traceback': is not a member of '_ts'
  C:\Users\Gudni\AppData\Local\Programs\Python\Python312\include\cpython/pystate.h(115): note: see declaration of '_ts'
  src/qoi/qoi.c(21988): error C2039: 'curexc_traceback': is not a member of '_ts'
  C:\Users\Gudni\AppData\Local\Programs\Python\Python312\include\cpython/pystate.h(115): note: see declaration of '_ts'
  src/qoi/qoi.c(23336): error C2039: 'ob_digit': is not a member of '_longobject'
  C:\Users\Gudni\AppData\Local\Programs\Python\Python312\include\cpython/longintrepr.h(87): note: see declaration of '_longobject'
  src/qoi/qoi.c(24904): error C2039: 'ob_digit': is not a member of '_longobject'
  C:\Users\Gudni\AppData\Local\Programs\Python\Python312\include\cpython/longintrepr.h(87): note: see declaration of '_longobject'
  src/qoi/qoi.c(24959): error C2039: 'ob_digit': is not a member of '_longobject'
  C:\Users\Gudni\AppData\Local\Programs\Python\Python312\include\cpython/longintrepr.h(87): note: see declaration of '_longobject'
  src/qoi/qoi.c(25100): error C2039: 'ob_digit': is not a member of '_longobject'
  C:\Users\Gudni\AppData\Local\Programs\Python\Python312\include\cpython/longintrepr.h(87): note: see declaration of '_longobject'
  src/qoi/qoi.c(25155): error C2039: 'ob_digit': is not a member of '_longobject'
  C:\Users\Gudni\AppData\Local\Programs\Python\Python312\include\cpython/longintrepr.h(87): note: see declaration of '_longobject'
  src/qoi/qoi.c(25296): error C2039: 'ob_digit': is not a member of '_longobject'
  C:\Users\Gudni\AppData\Local\Programs\Python\Python312\include\cpython/longintrepr.h(87): note: see declaration of '_longobject'
  src/qoi/qoi.c(25351): error C2039: 'ob_digit': is not a member of '_longobject'
  C:\Users\Gudni\AppData\Local\Programs\Python\Python312\include\cpython/longintrepr.h(87): note: see declaration of '_longobject'
  src/qoi/qoi.c(25530): error C2039: 'ob_digit': is not a member of '_longobject'
  C:\Users\Gudni\AppData\Local\Programs\Python\Python312\include\cpython/longintrepr.h(87): note: see declaration of '_longobject'
  src/qoi/qoi.c(25585): error C2039: 'ob_digit': is not a member of '_longobject'
  C:\Users\Gudni\AppData\Local\Programs\Python\Python312\include\cpython/longintrepr.h(87): note: see declaration of '_longobject'
  src/qoi/qoi.c(25936): error C2039: 'ob_digit': is not a member of '_longobject'
  C:\Users\Gudni\AppData\Local\Programs\Python\Python312\include\cpython/longintrepr.h(87): note: see declaration of '_longobject'
  error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2022\\BuildTools\\VC\\Tools\\MSVC\\14.37.32822\\bin\\HostX86\\x64\\cl.exe' failed with exit code 2
  [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for qoi
Failed to build qoi
ERROR: Could not build wheels for qoi, which is required to install pyproject.toml-based projects
```

@synodriver thanks for looking into this!

The bug you found, does it explain the build error I saw?

> error C2039: 'curexc_traceback': is not a member of '_ts'

Is it just a consequence of the malloc call failing? I’m wondering if you could reproduce the same error message as I was getting.

> @synodriver thanks for looking into this!
> 
> The bug you found, does it explain the build error I saw?
> 
> > error C2039: 'curexc_traceback': is not a member of '_ts'
> > error C2039: 'ob_digit': is not a member of '_longobject'
> 
> Is it just a consequence of the malloc call failing? I’m wondering if you could reproduce the same error message as I was getting.

I'll build on Debian again to see what happens

--- a/setup.py
+++ b/setup.py
@@ -32,7 +32,7 @@ extensions = [
     Extension(
         "qoi.qoi",
         sources=["src/qoi/qoi.pyx", "src/qoi/implementation.c"],
-        define_macros=[("QOI_MALLOC", "PyMem_Malloc"), ("QOI_FREE", "PyMem_Free"), ("NPY_NO_DEPRECATED_API", "NPY_1_7_API_VERSION")],
+        define_macros=[("QOI_MALLOC", "PyMem_RawMalloc"), ("QOI_FREE", "PyMem_RawFree"), ("NPY_NO_DEPRECATED_API", "NPY_1_7_API_VERSION")],
     )
 ]
 

--- a/src/qoi/qoi.pyx
+++ b/src/qoi/qoi.pyx
@@ -30,7 +30,7 @@ cdef class PixelWrapper:
 
 cpdef int write(filename, unsigned char[:,:,::1] rgb, colorspace: QOIColorSpace = QOIColorSpace.SRGB) except? -1:
     cdef bytes filename_bytes 
-    cdef char* _filename
+    cdef const char* _filename
     cdef qoi.qoi_desc desc
     cdef int ret
 
@@ -53,7 +53,7 @@ cpdef int write(filename, unsigned char[:,:,::1] rgb, colorspace: QOIColorSpace 
     
     cdef int bytes_written
     with nogil:
-        bytes_written = qoi.qoi_write(_filename, &rgb[0][0][0], &desc)
+        bytes_written = qoi.qoi_write(_filename, &rgb[0, 0, 0], &desc)
     if bytes_written == 0:
         raise RuntimeError("Failed to write!")
     return bytes_written
@@ -61,7 +61,7 @@ cpdef int write(filename, unsigned char[:,:,::1] rgb, colorspace: QOIColorSpace 
 cpdef np.ndarray[DTYPE_t, ndim=3] read(filename, int channels = 0, unsigned char[::1] colorspace = bytearray(1)):
     # TODO: how to return desc.colorspace? A: How about return a tuple of ndarray and a wrapper around struct qoi_desc? or we can add another param like char[:] to simulate pointer
     cdef bytes filename_bytes
-    cdef char* _filename
+    cdef const char* _filename
     cdef qoi.qoi_desc desc
     cdef int ret
     cdef char* pixels
@@ -99,7 +99,7 @@ cpdef bytes encode(unsigned char[:,:,::1] rgb, colorspace: QOIColorSpace = QOICo
     # if not rgb.flags['C_CONTIGUOUS']:
     #     rgb = np.ascontiguousarray(rgb) # makes a contiguous copy of the numpy array so we can read memory directly
     with nogil:
-        encoded = <char *>qoi.qoi_encode(&rgb[0][0][0], &desc, &size)
+        encoded = <char *>qoi.qoi_encode(&rgb[0, 0, 0], &desc, &size)
     if encoded is NULL or size <= 0:
         raise RuntimeError("Failed to encode!")
     try:

